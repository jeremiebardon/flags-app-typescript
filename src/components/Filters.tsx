import { CloseIcon } from "@chakra-ui/icons";
import { FormControl, Icon, Input, InputGroup, InputLeftElement, InputRightElement, Stack, Text } from "@chakra-ui/react";
import { useCallback, useEffect, useRef, useState } from "react";
import { FiSearch } from 'react-icons/fi';
import { useDebounce } from "../hooks/useDebounce.hook";
import { Region } from "../types/countries.type";
import { Option } from "./CustomSelect/Option";
import { CustomSelect } from "./CustomSelect/Select";

interface FiltersProps {
  handleNameChange: (value: string) => void;
  handleRegionChange: (value: Region) => void;
}

type FilterComponent = React.FunctionComponent<FiltersProps>

export const Filters: FilterComponent = ({ handleNameChange, handleRegionChange }: FiltersProps) => {
  const inputRef = useRef<HTMLInputElement>(null);

  const [nameValue, setNameValue] = useState<string>('');
  const debouncedValue = useDebounce<string>(nameValue, 400)

  const onNameChange = useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
    setNameValue(event.target.value)
  }, []);


  const resetName = () => {
    if (!!inputRef.current) {
      inputRef.current.value = '';
      setNameValue('');
    }
  };

  const onRegionChange = useCallback((region: Region) => {
    if (!!inputRef.current) {
      inputRef.current.value = '';
      setNameValue('');
    }
    
    handleRegionChange(region);
  }, [handleRegionChange]);

  useEffect(() => {
    handleNameChange(debouncedValue);
  }, [debouncedValue, handleNameChange])

  return (
    <Stack direction={['column', 'row']} justifyContent="space-between">
      <InputGroup maxW="xs" boxShadow='md' maxWidth={["100%", "30rem"]}>
        <InputLeftElement pointerEvents="none" height='100%'>
          <Icon as={FiSearch} color="gray.300" boxSize="4" />
        </InputLeftElement>
        <Input
          data-testid="input-name"
          size='lg'
          borderRadius={3}
          ref={inputRef}
          placeholder="Search for a country "
          onChange={(e: React.ChangeEvent<HTMLInputElement>) => onNameChange(e)}
          _placeholder={{ opacity: 1, color: 'gray.300' }}
          _focus={{ borderColor: 'none' }}
        />
        { nameValue.length > 0 && 
          <InputRightElement height='100%'>
            <Icon as={CloseIcon} color="gray.500" boxSize="4" onClick={resetName} _hover={{ cursor: 'pointer' }} data-testid="reset-name"/>
          </InputRightElement>
        }
      </InputGroup>

     <FormControl id="region" width={['100%', '12.5rem']}>
        <CustomSelect
          data-testid="select-region"
          name="region"
          size='lg'
          boxShadow='md'
          colorScheme="blue"
          onChange={(region) => onRegionChange(region as Region)}
          placeholder="Filter by Region"
        >
          <Option value='all'>
            <Text> All </Text>
          </Option>
          <Option value="africa">
            <Text> Africa </Text>
          </Option>
          <Option value="americas">
            <Text> Americas </Text>
          </Option>
          <Option value="asia">
            <Text> Asia </Text>
          </Option>
          <Option value="europe">
            <Text> Europe </Text>
          </Option>
          <Option value="oceania">
            <Text> Oceania </Text>
          </Option>
        </CustomSelect>
      </FormControl>
    </Stack>
  )
}