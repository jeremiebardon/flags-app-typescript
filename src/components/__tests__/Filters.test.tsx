import { fireEvent, render } from "@testing-library/react";
import { Filters } from "../Filters";




describe('<Filters />', () => {
  const handleNameChangeMock = jest.fn();
  const handleRegionChangeMock = jest.fn();
  
  it('should render', () => {
    const wrapper = render(<Filters handleNameChange={handleNameChangeMock} handleRegionChange={handleRegionChangeMock} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should change name and call function', () => {
    const { getByTestId } = render(<Filters handleNameChange={handleNameChangeMock} handleRegionChange={handleRegionChangeMock} />)

    const input = getByTestId('input-name');

    fireEvent.change(input, { target: { value: 'name' } });

    expect(handleNameChangeMock).toHaveBeenCalled();
  });

  it('should reset name search', () => {
    const { getByTestId } = render(<Filters handleNameChange={handleNameChangeMock} handleRegionChange={handleRegionChangeMock} />)
    const input = getByTestId('input-name');

    fireEvent.change(input, { target: { value: 'name' } });

    const inputReset = getByTestId('reset-name');

    fireEvent.click(inputReset);

    expect((input as HTMLInputElement).value).toEqual('');
  });
});