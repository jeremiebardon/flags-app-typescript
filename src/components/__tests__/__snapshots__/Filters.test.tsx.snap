// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filters /> should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="chakra-stack css-om1qj2"
      >
        <div
          class="chakra-input__group css-1gvctzy"
        >
          <div
            class="chakra-input__left-element css-o061fv"
          >
            <svg
              class="chakra-icon css-ykyi6j"
              fill="none"
              focusable="false"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="11"
                cy="11"
                r="8"
              />
              <line
                x1="21"
                x2="16.65"
                y1="21"
                y2="16.65"
              />
            </svg>
          </div>
          <input
            class="chakra-input css-1yiizff"
            data-testid="input-name"
            placeholder="Search for a country "
          />
        </div>
        <div
          class="chakra-form-control css-1t02n1q"
          role="group"
        >
          <div
            class="css-79elbk"
          >
            <button
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="downshift-0-label downshift-0-toggle-button"
              class="css-17hs7hi"
              data-focus-visible-added="false"
              data-testid="select-region"
              id="downshift-0-toggle-button"
              name="region"
            >
              <span
                class="css-188y514"
              >
                Filter by Region
              </span>
              <svg
                aria-hidden="true"
                fill="currentColor"
                focusable="false"
                role="presentation"
                style="width: 1em; height: 1em; color: currentColor;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <div
              class="css-1pye1st"
              style="position: absolute; min-width: max-content; inset: 0 auto auto 0; visibility: hidden;"
            >
              <ul
                aria-labelledby="downshift-0-label"
                class="css-0"
                data-focus-visible-added="false"
                id="downshift-0-menu"
                role="listbox"
                tabindex="-1"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-stack css-om1qj2"
    >
      <div
        class="chakra-input__group css-1gvctzy"
      >
        <div
          class="chakra-input__left-element css-o061fv"
        >
          <svg
            class="chakra-icon css-ykyi6j"
            fill="none"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="11"
              cy="11"
              r="8"
            />
            <line
              x1="21"
              x2="16.65"
              y1="21"
              y2="16.65"
            />
          </svg>
        </div>
        <input
          class="chakra-input css-1yiizff"
          data-testid="input-name"
          placeholder="Search for a country "
        />
      </div>
      <div
        class="chakra-form-control css-1t02n1q"
        role="group"
      >
        <div
          class="css-79elbk"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="downshift-0-label downshift-0-toggle-button"
            class="css-17hs7hi"
            data-focus-visible-added="false"
            data-testid="select-region"
            id="downshift-0-toggle-button"
            name="region"
          >
            <span
              class="css-188y514"
            >
              Filter by Region
            </span>
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              role="presentation"
              style="width: 1em; height: 1em; color: currentColor;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </button>
          <div
            class="css-1pye1st"
            style="position: absolute; min-width: max-content; inset: 0 auto auto 0; visibility: hidden;"
          >
            <ul
              aria-labelledby="downshift-0-label"
              class="css-0"
              data-focus-visible-added="false"
              id="downshift-0-menu"
              role="listbox"
              tabindex="-1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
